File: mcp/juno_mcp/integrity.py
"""Tool-definition integrity for our own MCP surface.

A tool description is not documentation. The model reads it as instruction, so
whoever controls it steers the agent. The two attacks that follow are named in
docs/threat-landscape.md §2.4:

* **Tool poisoning** — instructions hidden in a description the user never reads.
* **Rug pull** — a server approved while benign later redefines its tools and
  inherits the approval. Most clients do not alert on the change.

We are an MCP surface, so both apply to us. This module pins what we serve.
Every tool's name, description and schema is hashed into `tools.lock.json`,
which is committed and reviewable; the server verifies itself against that lock
before it will answer anything.

What that buys, stated honestly:

* A dependency update, a tampered install, or an edit that reaches the package
  after review cannot silently change what the agent is told — the server
  refuses to start and says which tool moved.
* The lock is a git-tracked artefact, so re-pinning is a diff somebody has to
  approve. That is the trust root: not this file, the review.

What it does not buy: an attacker who can edit the source *and* the lock in the
same reviewed commit is not stopped by a hash. That is a code-review problem,
and the lock is what makes it visible in one.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

LOCK_PATH = Path(__file__).with_name("tools.lock.json")

LOCK_VERSION = 1
ALGORITHM = "sha256"

# Everything the model or the client can be steered by. `icons`, `meta` and
# `execution` are deliberately excluded: they are presentation and transport
# detail that the SDK may set differently between versions, and pinning them
# would produce noisy failures that teach people to re-pin without reading.
SIGNIFICANT_FIELDS = (
    "name",
    "title",
    "description",
    "input_schema",
    "output_schema",
    "annotations",
)


def _as_dict(tool: Any) -> dict[str, Any]:
    if hasattr(tool, "model_dump"):
        return tool.model_dump(mode="json")
    if isinstance(tool, dict):
        return dict(tool)
    raise TypeError(f"cannot read tool definition from {type(tool)!r}")


def canonical(tool: Any) -> dict[str, Any]:
    """The part of a tool definition that can change agent behaviour."""
    raw = _as_dict(tool)
    # Accept either snake_case (SDK objects) or camelCase (wire JSON).
    aliases = {"input_schema": "inputSchema", "output_schema": "outputSchema"}
    out: dict[str, Any] = {}
    for field_name in SIGNIFICANT_FIELDS:
        value = raw.get(field_name)
        if value is None and field_name in aliases:
            value = raw.get(aliases[field_name])
        out[field_name] = value
    return out


def _encode(payload: Any) -> bytes:
    return json.dumps(
        payload, sort_keys=True, separators=(",", ":"), ensure_ascii=False
    ).encode("utf-8")


def digest(tool: Any) -> str:
    return hashlib.sha256(_encode(canonical(tool))).hexdigest()


def build_manifest(tools: list[Any]) -> dict[str, Any]:
    """A per-tool digest plus one digest over the whole surface.

    The combined digest covers the tool *set*, so adding or removing a tool is a
    change even though every surviving tool still matches.
    """
    per_tool = {}
    for tool in tools:
        entry = canonical(tool)
        name = entry["name"]
        if name in per_tool:
            raise ValueError(f"duplicate tool name: {name}")
        per_tool[name] = digest(tool)

    combined = hashlib.sha256(_encode(dict(sorted(per_tool.items())))).hexdigest()
    return {
        "lock_version": LOCK_VERSION,
        "algorithm": ALGORITHM,
        "tools": dict(sorted(per_tool.items())),
        "surface": combined,
    }


@dataclass
class Report:
    """What changed, in the terms an operator needs to decide if it is fine."""

    ok: bool
    reason: str = ""
    added: list[str] = field(default_factory=list)
    removed: list[str] = field(default_factory=list)
    changed: list[str] = field(default_factory=list)

    def summary(self) -> str:
        if self.ok:
            return "tool definitions match tools.lock.json"
        parts = []
        if self.added:
            parts.append(f"added: {', '.join(sorted(self.added))}")
        if self.removed:
            parts.append(f"removed: {', '.join(sorted(self.removed))}")
        if self.changed:
            parts.append(f"redefined: {', '.join(sorted(self.changed))}")
        detail = "; ".join(parts)
        return f"{self.reason}{' — ' if self.reason and detail else ''}{detail}"


def load_lock(path: Path = LOCK_PATH) -> dict[str, Any] | None:
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except FileNotFoundError:
        return None
    except (OSError, json.JSONDecodeError):
        return {}


def verify(tools: list[Any], lock: dict[str, Any] | None) -> Report:
    """Compare the live surface against the pinned one.

    A missing or unreadable lock is a failure, not a pass. "We could not check"
    is the same answer as "it does not match" everywhere else in this product.
    """
    if lock is None:
        return Report(False, "tools.lock.json is missing")
    if not lock or "tools" not in lock:
        return Report(False, "tools.lock.json is unreadable")
    if lock.get("lock_version") != LOCK_VERSION:
        return Report(
            False, f"tools.lock.json is version {lock.get('lock_version')!r}, expected {LOCK_VERSION}"
        )
    if lock.get("algorithm") != ALGORITHM:
        return Report(False, f"unsupported digest algorithm {lock.get('algorithm')!r}")

    current = build_manifest(tools)
    pinned_tools: dict[str, str] = lock["tools"]
    live_tools: dict[str, str] = current["tools"]

    added = sorted(set(live_tools) - set(pinned_tools))
    removed = sorted(set(pinned_tools) - set(live_tools))
    changed = sorted(
        name
        for name in set(live_tools) & set(pinned_tools)
        if live_tools[name] != pinned_tools[name]
    )

    if added or removed or changed:
        return Report(
            False,
            "the tool surface does not match the reviewed definitions",
            added=added,
            removed=removed,
            changed=changed,
        )

    # Belt and braces: the set matched tool by tool, so this can only disagree if
    # the lock's own combined digest was edited by hand.
    if lock.get("surface") != current["surface"]:
        return Report(False, "tools.lock.json surface digest does not match its own entries")

    return Report(True)


# Development escape hatch. Deliberately an environment variable rather than a
# config file setting: serving an unverified tool surface should take a
# conscious act at the command line, and leave a trace in the shell history.
UNPINNED_ENV = "JUNO_MCP_ALLOW_UNPINNED"

# EX_CONFIG. A refusal to start is the alert — the MCP client shows the server
# red, which is louder than a log line nobody reads.
REFUSE_EXIT_CODE = 78


@dataclass
class Decision:
    """Whether to serve, and what to tell the operator either way."""

    serve: bool
    message: str = ""


def gate(report: Report, allow_unpinned: bool = False) -> Decision:
    """Turn a verification result into a serve-or-refuse decision.

    Kept here, away from the SDK, so the policy is testable without an MCP
    runtime — the reason this module imports nothing but the standard library.
    """
    if report.ok:
        return Decision(True)

    if allow_unpinned:
        return Decision(
            True,
            f"juno-mcp: WARNING — {report.summary()}. "
            f"Serving anyway because {UNPINNED_ENV} is set.",
        )

    return Decision(
        False,
        f"juno-mcp: refusing to start — {report.summary()}.\n"
        f"  The tool descriptions an agent reads are instructions, so a change here "
        f"is a change to agent behaviour.\n"
        f"  Review the diff. If it is intended, re-pin with: "
        f"python -m juno_mcp --update-lock\n"
        f"  To serve unverified anyway (development only): {UNPINNED_ENV}=1",
    )


def write_lock(tools: list[Any], path: Path = LOCK_PATH) -> dict[str, Any]:
    manifest = build_manifest(tools)
    path.write_text(json.dumps(manifest, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    return manifest

File: packages/junoguard/src/integrity.ts
/**
 * Tool-definition integrity for the published MCP surface.
 *
 * A tool description is not documentation. The model reads it as instruction,
 * so whoever controls it steers the agent. Two attacks follow, both named in
 * docs/threat-landscape.md §2.4:
 *
 * - **Tool poisoning** — instructions hidden in a description no human reads.
 * - **Rug pull** — a server approved while benign later redefines its tools and
 *   keeps the approval. Most clients never alert on the change.
 *
 * This is the package an agent actually runs, so both apply to us directly.
 * Every tool's name, description and schema is hashed into `tools.lock.json`,
 * which is committed and published; the server verifies itself against that
 * lock before it serves anything.
 *
 * What that buys, stated honestly:
 *
 * - A tampered install, a compromised dependency, or an edit that lands after
 *   review cannot silently change what the agent is told. The server refuses to
 *   start and names the tool that moved.
 * - The lock is a git-tracked artefact, so re-pinning is a diff somebody has to
 *   approve. That is the trust root: not this file, the review.
 *
 * What it does not buy: an attacker who edits the source *and* the lock in one
 * reviewed commit is not stopped by a hash. The lock is what makes that visible
 * in a diff.
 */

import { createHash } from "node:crypto";
import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

/** Resolves to the package root from both `src/` (tsx) and `dist/` (published). */
export const LOCK_PATH = fileURLToPath(new URL("../tools.lock.json", import.meta.url));

export const LOCK_VERSION = 1;
export const ALGORITHM = "sha256";

/** Serving an unverified surface should take a conscious act at the command line. */
export const UNPINNED_ENV = "JUNO_MCP_ALLOW_UNPINNED";

/**
 * EX_CONFIG. Refusing to start is the alert — the client shows the server red,
 * which is louder than a log line nobody reads.
 */
export const REFUSE_EXIT_CODE = 78;

/**
 * Everything the model or the client can be steered by. Presentation and
 * transport detail the SDK may set differently between versions is excluded on
 * purpose: pinning it produces noisy failures, and noisy failures teach people
 * to re-pin without reading.
 */
const SIGNIFICANT_FIELDS = [
  "name",
  "title",
  "description",
  "inputSchema",
  "outputSchema",
  "annotations",
] as const;

export interface ToolLike {
  name: string;
  [key: string]: unknown;
}

export interface Manifest {
  lock_version: number;
  algorithm: string;
  tools: Record<string, string>;
  surface: string;
}

export interface Report {
  ok: boolean;
  reason: string;
  added: string[];
  removed: string[];
  changed: string[];
}

export interface Decision {
  serve: boolean;
  message: string;
}

/** Stable JSON: sorted keys at every depth, so digests do not depend on key order. */
function stableStringify(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value) ?? "null";
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(",")}]`;
  const entries = Object.keys(value as Record<string, unknown>)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${stableStringify((value as Record<string, unknown>)[key])}`);
  return `{${entries.join(",")}}`;
}

export function canonical(tool: ToolLike): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const field of SIGNIFICANT_FIELDS) {
    out[field] = tool[field] ?? null;
  }
  return out;
}

export function digest(tool: ToolLike): string {
  return createHash("sha256").update(stableStringify(canonical(tool)), "utf8").digest("hex");
}

export function buildManifest(tools: ToolLike[]): Manifest {
  const perTool: Record<string, string> = {};
  for (const tool of tools) {
    if (tool.name in perTool) throw new Error(`duplicate tool name: ${tool.name}`);
    perTool[tool.name] = digest(tool);
  }

  const sorted = Object.fromEntries(Object.entries(perTool).sort(([a], [b]) => (a < b ? -1 : 1)));
  // One digest over the whole surface, so adding or removing a tool is a change
  // even though every surviving tool still matches.
  const surface = createHash("sha256").update(stableStringify(sorted), "utf8").digest("hex");

  return { lock_version: LOCK_VERSION, algorithm: ALGORITHM, tools: sorted, surface };
}

export function summarise(report: Report): string {
  if (report.ok) return "tool definitions match tools.lock.json";
  const parts: string[] = [];
  if (report.added.length) parts.push(`added: ${[...report.added].sort().join(", ")}`);
  if (report.removed.length) parts.push(`removed: ${[...report.removed].sort().join(", ")}`);
  if (report.changed.length) parts.push(`redefined: ${[...report.changed].sort().join(", ")}`);
  const detail = parts.join("; ");
  return report.reason && detail ? `${report.reason} — ${detail}` : report.reason || detail;
}

function fail(reason: string, extra: Partial<Report> = {}): Report {
  return { ok: false, reason, added: [], removed: [], changed: [], ...extra };
}

export function loadLock(path: string = LOCK_PATH): Manifest | null | undefined {
  let raw: string;
  try {
    raw = readFileSync(path, "utf8");
  } catch {
    return null; // missing
  }
  try {
    return JSON.parse(raw) as Manifest;
  } catch {
    return undefined; // present but unreadable
  }
}

/**
 * Compare the live surface against the pinned one. A missing or unreadable lock
 * is a failure, not a pass: "we could not check" is the same answer as "it does
 * not match" everywhere else in this product.
 */
export function verify(tools: ToolLike[], lock: Manifest | null | undefined): Report {
  if (lock === null) return fail("tools.lock.json is missing");
  if (!lock || typeof lock !== "object" || !lock.tools) return fail("tools.lock.json is unreadable");
  if (lock.lock_version !== LOCK_VERSION) {
    return fail(`tools.lock.json is version ${lock.lock_version}, expected ${LOCK_VERSION}`);
  }
  if (lock.algorithm !== ALGORITHM) return fail(`unsupported digest algorithm ${lock.algorithm}`);

  const current = buildManifest(tools);
  const live = current.tools;
  const pinned = lock.tools;

  const added = Object.keys(live).filter((n) => !(n in pinned)).sort();
  const removed = Object.keys(pinned).filter((n) => !(n in live)).sort();
  const changed = Object.keys(live)
    .filter((n) => n in pinned && live[n] !== pinned[n])
    .sort();

  if (added.length || removed.length || changed.length) {
    return fail("the tool surface does not match the reviewed definitions", {
      added,
      removed,
      changed,
    });
  }

  // Belt and braces: matching tool by tool, this can only disagree if the lock's
  // own combined digest was edited by hand.
  if (lock.surface !== current.surface) {
    return fail("tools.lock.json surface digest does not match its own entries");
  }

  return { ok: true, reason: "", added: [], removed: [], changed: [] };
}

/**
 * Turn a verification result into a serve-or-refuse decision. Kept separate
 * from the SDK so the policy is testable without an MCP runtime.
 */
export function gate(report: Report, allowUnpinned = false): Decision {
  if (report.ok) return { serve: true, message: "" };

  if (allowUnpinned) {
    return {
      serve: true,
      message:
        `junoguard: WARNING — ${summarise(report)}. ` +
        `Serving anyway because ${UNPINNED_ENV} is set.`,
    };
  }

  return {
    serve: false,
    message:
      `junoguard: refusing to start — ${summarise(report)}.\n` +
      `  The tool descriptions an agent reads are instructions, so a change here ` +
      `is a change to agent behaviour.\n` +
      `  Review the diff. If it is intended, re-pin with: juno mcp --update-lock\n` +
      `  To serve unverified anyway (development only): ${UNPINNED_ENV}=1`,
  };
}

export function allowUnpinnedFromEnv(env: NodeJS.ProcessEnv = process.env): boolean {
  return ["1", "true", "yes"].includes((env[UNPINNED_ENV] ?? "").trim().toLowerCase());
}

export function writeLock(tools: ToolLike[], path: string = LOCK_PATH): Manifest {
  const manifest = buildManifest(tools);
  writeFileSync(path, `${JSON.stringify(manifest, null, 2)}\n`, "utf8");
  return manifest;
}
